### PR TITLE
PETSc - add MPI_Reduce and PetscBarrier for benchmarking

### PR DESCRIPTION
**Pull request by jeremylt**
This fixes a small bug in the PETSc BPs when running in parallel.

### CODE REVIEW

File: examples/petsc/bps.c [hunk 1]
```diff
@@ -762,6 +762,7 @@ int main(int argc, char **argv) {
     my_rt_start = MPI_Wtime();
     ierr = KSPSolve(ksp, rhs, X); CHKERRQ(ierr);
     my_rt = MPI_Wtime() - my_rt_start;
+    MPI_Reduce(&my_rt, &my_rt, 1, MPI_DOUBLE, MPI_MIN, 0, comm);
```
Comment (jedbrown) [inline examples/petsc/bps.c]:
```suggestion
    MPI_Allreduce(&my_rt, &my_rt, 1, MPI_DOUBLE, MPI_MIN, comm);
```

File: examples/petsc/bps.c [hunk 2]
```diff
@@ -772,6 +773,7 @@ int main(int argc, char **argv) {
     }
   }
   // Timed solve
+  ierr = PetscBarrier(NULL); CHKERRQ(ierr);
```
Comment (jedbrown) [inline examples/petsc/bps.c]:
```suggestion
  ierr = PetscBarrier((PetscObject)ksp); CHKERRQ(ierr);
```

File: examples/petsc/bpsdmplex.c [hunk 3]
```diff
@@ -674,6 +674,7 @@ int main(int argc, char **argv) {
     my_rt_start = MPI_Wtime();
     ierr = KSPSolve(ksp, rhs, X); CHKERRQ(ierr);
     my_rt = MPI_Wtime() - my_rt_start;
+    MPI_Reduce(&my_rt, &my_rt, 1, MPI_DOUBLE, MPI_MIN, 0, comm);
```
Comment (jedbrown) [inline examples/petsc/bpsdmplex.c]:
```suggestion
    MPI_Allreduce(&my_rt, &my_rt, 1, MPI_DOUBLE, MPI_MIN, comm);
```

File: examples/petsc/bpsdmplex.c [hunk 4]
```diff
@@ -685,6 +686,7 @@ int main(int argc, char **argv) {
   }
 
   // Timed solve
+  ierr = PetscBarrier(NULL); CHKERRQ(ierr);
```
Comment (jedbrown) [inline examples/petsc/bpsdmplex.c]:
```suggestion
  ierr = PetscBarrier((PetscObject)ksp); CHKERRQ(ierr);
```

File: examples/navier-stokes/Makefile [hunk 5]
```diff
@@ -14,7 +14,7 @@
 # software, applications, hardware, advanced system engineering and early
 # testbed platforms, in support of the nation's exascale computing imperative.
 
-OPT ?= -O -g -march=native -ffp-contract=fast -fopenmp-simd
+OPT ?= -O3 -g -march=native -ffp-contract=fast -fopenmp-simd
```
Comment (jedbrown) [inline examples/navier-stokes/Makefile]: Is this intentional? (I don't object, but it's unclear whether you wanted to add it.)

File: examples/petsc/bps.c [hunk 6]
```diff
@@ -762,6 +762,7 @@ int main(int argc, char **argv) {
     my_rt_start = MPI_Wtime();
     ierr = KSPSolve(ksp, rhs, X); CHKERRQ(ierr);
     my_rt = MPI_Wtime() - my_rt_start;
+    MPI_Allreduce(&my_rt, &my_rt, 1, MPI_DOUBLE, MPI_MIN, comm);
```
Comment (jedbrown) [inline examples/petsc/bps.c]: Oops, should have been.
```suggestion
    MPI_Allreduce(MPI_IN_PLACE, &my_rt, 1, MPI_DOUBLE, MPI_MIN, comm);
```

File: examples/petsc/bpsdmplex.c [hunk 7]
```diff
@@ -674,6 +674,7 @@ int main(int argc, char **argv) {
     my_rt_start = MPI_Wtime();
     ierr = KSPSolve(ksp, rhs, X); CHKERRQ(ierr);
     my_rt = MPI_Wtime() - my_rt_start;
+    MPI_Allreduce(&my_rt, &my_rt, 1, MPI_DOUBLE, MPI_MIN, comm);
```
Comment (jedbrown) [inline examples/petsc/bpsdmplex.c]:
```suggestion
    MPI_Allreduce(MPI_IN_PLACE, &my_rt, 1, MPI_DOUBLE, MPI_MIN, comm);
```

**Comment by jeremylt** [comment]
@jedbrown, should we change the MPI_Reduce on line 802/803 to an MPI_Allreduce as well?

**Comment by jedbrown** [comment]
I would, yes.  (Though it won't actually be used on anything but rank 0 -- `PetscPrintf` skips output from all but rank 0.)

**Comment by jeremylt** [comment]
Sounds good. I'll squash+merge once Travis finishes.

**Pull request merged**